c10/util/Float8_e5m2.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+/// Defines the Float8_e5m2 type (8-bit floating-point) including conversions
+/// to standard C types and basic arithmetic operations. Note that arithmetic
+/// operations are implemented by converting to floating point and
+/// performing the operation in float32.
+/// Binary configuration:
+/// s eeeee mm
+/// 1 sign bit
+/// 5 exponent bits
+/// 2 mantissa bits
+/// bias = 15
+///
+/// Implementation based on the paper https://arxiv.org/pdf/2209.05433.pdf
+/// and inspired by Half implementation from pytorch/c10/util/Half.h
+
+#include <c10/util/Half.h>
+
+namespace c10 {
+
+namespace detail {
+
+/*
+ * Convert a 8-bit floating-point number in fp8 E5M2 format, in bit
+ * representation, to a 32-bit floating-point number in IEEE single-precision
+ * format, in bit representation.
+ *
+ * @note The implementation doesn't use any floating-point operations.
+ */
+inline C10_HOST_DEVICE float fp8e5m2_to_fp32_value(uint8_t input) {
+  /*
+   * Extend the fp8 E5M2 number to 32 bits and shift to the
+   * upper part of the 32-bit word:
+   *      +---+----+---+-----------------------------+
+   *      | S |EEEEE|MM|0000 0000 0000 0000 0000 0000|
+   *      +---+----+---+-----------------------------+
+   * Bits  31 26-30 24-25          0-23
+   *
+   * S - sign bit, E - bits of the biased exponent, M - bits of the mantissa, 0
+   * - zero bits.
+   */
+  uint16_t half_representation = input;
+  half_representation <<= 8;
+  return fp16_ieee_to_fp32_value(half_representation);
+}
+
+/*
+ * Convert a 32-bit floating-point number in IEEE single-precision format to a
+ * 8-bit floating-point number in fp8 E5M2 format, in bit representation.
+ */
+inline C10_HOST_DEVICE uint8_t fp8e5m2_from_fp32_value(float f) {
+  /*
+   * Binary representation of fp32 infinity
+   * 0 11111111 00000000000000000000000
+   */
+  constexpr uint32_t fp32_inf = UINT32_C(255) << 23;
+
+  /*
+   * Binary representation of 65536.0f, which is the first value
+   * not representable in fp8e5m2 range:
+   * 0 11111 00 - fp8e5m2
+   * 0 10001111 00000000000000000000000 - fp32
+   */
+  constexpr uint32_t fp8_max = UINT32_C(143) << 23;
+
+  /*
+   * A mask for converting fp32 numbers lower than fp8e5m2 normal range
+   * into denorm representation
+   * magic number: ((127 - 15) + (23 - 2) + 1)
+   */
+  constexpr uint32_t denorm_mask = UINT32_C(134) << 23;
+
+  uint32_t f_bits = fp32_to_bits(f);
+  uint8_t result = 0u;
+
+  /*
+   * Extract the sign of the input number into the high bit of the 32-bit word:
+   *
+   *      +---+----------------------------------+
+   *      | S |0000000 00000000 00000000 00000000|
+   *      +---+----------------------------------+
+   * Bits  31                 0-31
+   */
+  const uint32_t sign = f_bits & UINT32_C(0x80000000);
+
+  /*
+   * Set sign bit to 0
+   */
+  f_bits ^= sign;
+
+  if (f_bits >= fp8_max) {
+    // NaN - all exponent and mantissa bits set to 1
+    result = f_bits > fp32_inf ? UINT8_C(0x7F) : UINT8_C(0x7C);
```

Comment:
Can I ask if PyTorch is only implementing the non-saturation mode for casts, as specified [here ](https://onnx.ai/onnx/technical/float8.html#cast)? 

The current default behavior for ONNX is to saturate: https://github.com/onnx/onnx/blob/main/docs/Operators.md#attributes-7.

Reply:
You're right, currently it implements only non-saturation mode. And, in consequence, is not aligned with float8 implementation from FBGEMM https://github.com/pytorch/FBGEMM/blob/main/fbgemm_gpu/include/fbgemm_gpu/quantize_ops_utils.h

@vkuzo, @albanD, was it intended to have saturation enabled by default in casting to float8 dtypes? If so, it needs to be added, because currently values out of float8 range are being casted to inf/nan. I can do it next week after vacation.

Also, should it be configurable via parameter of .to() operator or some other way?

Reply:
Thanks for the comment!  During PR review we punted these questions to future PRs just to get the initial dtype in, I guess now it's the time to discuss them :)

I would make an educated guess that saturating by default would make sense if people are doing delayed scaling, but let me think on that / talk to some of the research teams using fp8.  

For if/how to make this configurable...let me also ask around on what would make sense.

Reply:
In terms of consistency, when given integer inputs, we raise an error for too large numbers.
For floating point inputs, all our dtypes will return inf.

So I think inf is the right call here.

I think that `a_half = a.clamp(max=torch.finfo(torch.half).max).to(dtype=torch.half)` is a good way to get the behavior you want if needed.

Reply:
Thanks for the information. 

Let me also add that when we introduced FP8 into ONNX one of the motivations for having saturation mode as the default was that during inference we don't want to end up with a lot of NaN's and Inf's propagating through the network.

cc: @nzmora @xadupre

Reply:
> Agreed that this might change when we have a stable decision on how to handle delayed scaling.
In the meantime keeping a behavior consistent with our other floating point dtypes (as done in this PR) sounds like the right way to go.

That sounds reasonable to me.  We'll just have to figure out how to expose the saturation option.  Ideally there should be a way to do to the saturated casts in eager mode without requiring a compiler pass.

Reply:
@vkuzo ,
Is there an open discussion about saturation mode?

Reply:
Here is the current plan:
1. the default will stay unsaturated as is, to align with @albanD's thoughts on matching default behavior of other floating point dtypes
2. @malfet is working on a private API to expose the saturated casts, so that we can bind to hardware accelerated saturated casts without requiring graph capture -> graph pass.  Name of this is TBD.
3. long term, we may revisit this after we get more data on how people use these dtypes

Reply:
It will be exposed to Python with a name preceded by an underscore, but there will initially be no BC guarantees.

Reply:
Thank you for the information.

Reply:
Hi @vkuzo @albanD @australopitek ,

May I know the current status of enabling fp8 delayed scaling and saturation in PyTorch?

For fp8 delayed scaling, it is a per-tensor one. So, is the implementation similar with Quantized operators? It cannot just use CPU dispatch, right? Because scaling factor need to be stored in Tensor. Please correct me if anything wrong.

And you mentioned that fp8 training with delayed scaling, what about fp8 inference? In https://github.com/NVIDIA/TransformerEngine, fp8 inference is also with delayed scaling as training. Will you take this into consideration of fp8 enabling?

Thanks so much! Hope to hear from you guys!

cc @mingfeima

Reply:
Hi @yanbing-j , please feel free to message @gchanan on PyTorch slack for context on your question.  I am on parental leave until early December but Greg can help on this while I'm out.